### ISSUE TITLE
Media Stream Source development

### ISSUE THREAD

**Issue by brabebhin** [Alovchin91/opus-winrt#1]
Hey there.
Sorry i've been silent these past few days. Been busy.

How are ya?
How's it going?

**Comment by Alovchin91** [comment]
@mcosmin222 Hey! It's going pretty good so far, I have compiled the original opus and opusfile libs for Windows Phone, and now I'm working on Windows Runtime wrapper. It would be almost the same as for Vorbis decoder, so I'm thinking about unifying an API scheme for all the three currently developed codecs, so I'm also making some changes to FLAC and Vorbis now.

**Comment by brabebhin** [comment]
Interesting topic here.

https://social.msdn.microsoft.com/Forums/en-US/wpdevelop/thread/7d312fc9-99ca-479d-b5ff-3c746c81091b/

I am kinda confused by the official response from Microsoft.

**Comment by Alovchin91** [comment]
@mcosmin222 Yes, it's confusing, but actually I think James is incorrect because the [documentation for MediaStreamSource](http://msdn.microsoft.com/library/windows/apps/dn282716) says this:

> **MediaStreamSource** allows apps to send compressed or uncompressed audio and video samples to the media pipeline for playback, transcoding, and streaming. Media samples can be dynamically generated by the app, or de-multiplexed from a stream or files. This flexibility enables apps to more easily extend platform support for new media formats or solve complex problems, such as adaptive streaming.

By the way, I answered you in that topic, maybe it will help :smiley:

**Comment by Alovchin91** [comment]
By the way, there is a [Media Extensions SDK](https://code.msdn.microsoft.com/windowsapps/Media-extensions-sample-7b466096/) that can be used to create a custom media extension by using native Media Foundation APIs, and then this extension can be registered by [MediaExtensionManager](http://msdn.microsoft.com/en-us/library/windows/apps/windows.media.mediaextensionmanager.aspx) to do all the decoding (and this way you would not need to use MediaStreamSource), but this is EXTREMELY complicated.

**Comment by Alovchin91** [comment]
I think we should consider developing MediaStreamSource adapter in C++/CX as well.

**Comment by brabebhin** [comment]
I think we are fine with MediaStreamSource.
As for writing in C++/CX: I suppose we can do that, but I don't think there will be any actual advantage. Using C# in winRT doesn't bring any dangers unless we use .net types (which is something one can easily do if they aren't paying attention). We are only using winRT types, so we are fine. In fact, using .net types by itself is not a danger, but rather converting between managed and native.

**Comment by Alovchin91** [comment]
@mcosmin Implementing MSS adapter in C++/CX will eliminate the marshalling because there will be no need to cross ABI boundaries :wink:

**Comment by brabebhin** [comment]
There is no marshaling done in our C# media stream sources. Even if the types appear like they are .net types, they actually reside in the unmanaged heap. Since we do not use any .net types, there is no need for marshaling to occur. Marshalling occurs when you convert IBuffer to Byte[] or the other way around. This is the trick of windows runtime: managed types appear unmanaged and the other way around. 

The reference held by the CLR will only be a simple pointer. Unless you do things with the values held by the reference pointers, you won't do any marshaling.

**Comment by Alovchin91** [comment]
@mcosmin222 Yeah, WinRT unmanaged objects do reside in unmanaged heap, but there are objects like strings (HSTRING in WinRT, System.String in .NET), arrays and maybe even basic value types that _may_ be marshalled in order to use the semantics of .NET. I say _may_ because I haven't seen the actual implementation of WinRT/.NET marshalling behavior, but I'm pretty confident that at least strings are marshalled. While this may not be the problem for streaming, it may create a delay before starting file playback (and we actually experience this delay). Moreover, since the adapter and .NET decoder glue are managed classes, they introduce some additional resource pressure. For example, after I removed `WaveInputStream` from FLAC's examples, the memory consumption of background playback process dropped from 25 MB to 20~21 MB, and this is pretty significant. I think we should give it a try.

**Comment by brabebhin** [comment]
Ok. I will implement them in CX.

**Comment by Alovchin91** [comment]
@mcosmin222 I think we could even implement `IMediaSource` using C++/CX :smiley: I'm not sure why we would need this but we could return `MediaStreamSource` instance from `IMFGetService.GetService` method :wink: This way we will make sure BackgroundAudioPlayer holds our adapter's reference for all the time of playback.

**Comment by brabebhin** [comment]
Let's not over-complicate ourselves. As long as it runs fine, we are fine. Regardless of language.

**Comment by brabebhin** [comment]
I have the flac CX media stream source almost ready. Gotta fix some bug with sample duration and timeline xD

**Comment by Alovchin91** [comment]
Okay, so I opened a Development discussion (#2) as the source code for the project has been commited.
I have noticed a strange memory consumption behavior of Opus' MediaStreamSource, so I'm going to check this one on the first place.
Please check if seeking works.

**Comment by brabebhin** [comment]
Where exactly did you hide the media stream source?

**Comment by Alovchin91** [comment]
@mcosmin222 I haven't posted it yet...

**Comment by brabebhin** [comment]
Well, I am in a bit of a pickle creating one myself. I can't seem to find the very important sampling rate for this format xD

**Comment by Alovchin91** [comment]
@mcosmin222 Sampling rate of Opus format is [always 48 kHz](http://opus-codec.org/docs/opusfile_api-0.6/index.html). If you need bits per sample value, it's 16 bits per sample ;)

**Comment by brabebhin** [comment]
OK...hmm...i suspected as much but I wasn't sure...about the sampling rate.

**Comment by brabebhin** [comment]
Anyway, I shall see this through tomorrow. 
wait, it is already tomorrow xD
